mm2src/lp_ordermatch/simple_market_maker.rs

```diff
@@ -440,30 +525,58 @@ async fn prepare_order(
         .await?
         .ok_or_else(|| MmError::new(OrderProcessingError::AssetNotEnabled))?;
 
-    info!("balance for {} is {}", cfg.base, base_balance);
+    debug!("balance for {} is {}", cfg.base, base_balance);
 
     let mut calculated_price = rates.price * cfg.spread.clone();
-    info!("calculated price is: {}", calculated_price);
+    debug!("calculated price is: {}", calculated_price);
     if cfg.check_last_bidirectional_trade_thresh_hold.unwrap_or(false) {
-        calculated_price = vwap_calculator(calculated_price.clone(), ctx, &cfg).await?;
+        calculated_price = vwap_calculator(calculated_price.clone(), ctx, cfg).await?;
+    }
+
+    let mut is_max = cfg.max.unwrap_or(false);
+
+    if cfg.balance_percent.is_some() && cfg.max_volume_usd.is_some() {
```

Comment:
```rust
enum VolumeSettings {
    BalancePct { min, max },
    Usd { min, max }
}
```
Will look much better and easier to handle and read than these `is_some`, `if let Some(_)` constructions.

Reply:
I Will do that, it's a good idea, I just need to find a way to deserialize all those stuff correctly with serde and then I can use pattern matching!

Reply:
We typically use adjacently tagged representation: https://serde.rs/enum-representations.html#adjacently-tagged

Reply:
Oh ok I was initially thinking with https://serde.rs/attr-flatten.html#struct-flattening so it's will flatten and I can choose the field name for each field of the enum. Will investigate anyway.

Reply:
> ```rust
> enum VolumeSettings {
>     BalancePct { min, max },
>     Usd { min, max }
> }
> ```
> 
> Will look much better and easier to handle and read than these `is_some`, `if let Some(_)` constructions.

I made some test locally with those settings but there is few things that are more painful:

- We can perfectly use a `min_vol_percentage` and a `max_volume_usd`
- min and max are always optional
- Code will still look painful:

```rust
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "type")]
enum VolumeSettings {
    #[serde(rename = "balance_percentage")]
    BalancePercentage {
        min: Option<MmNumber>,
        percent: Option<MmNumber>,
    },

    #[serde(rename = "usd")]
    Usd {
        min: Option<MmNumber>,
        max: Option<MmNumber>,
    },
}

let volume = if let Some(volume_settings) = &cfg.volume_settings {
        match volume_settings {
            VolumeSettings::BalancePercentage { percent, .. } => {
                if let Some(balance_percent) = percent {
                    if *balance_percent > MmNumber::from(1) {
                        is_max = true;
                        MmNumber::default()
                    } else {
                        balance_percent * &base_balance
                    }
                }
            },
            VolumeSettings::Usd { max, .. } => {
                if let Some(max_volume_usd) = max {
                    if &base_balance * &rates.base_price < *max_volume_usd {
                        is_max = true;
                        MmNumber::default()
                    } else {
                        max_volume_usd / &rates.base_price
                    }
                }
            },
        }
    } else {
        MmNumber::default()
    };
  ```
  
  Even with this approach, it makes everything more complex and less flexible than the current situation - any hints?

Reply:
I will try with a structure shortly something different.

Reply:
I see. I misunderstood first that `We can optionally use a min_vol_percentage and a max_volume_usd (with this new approach is not possible)` is a valid situation. What do you think about a more specific enum as follows?
```rust
enum BotVolumeOption {
    BalancePct(MmNumber),
    Usd(MmNumber),
}

struct VolumeSettings {
    min_volume: BotVolumeOption,
    max_volume: BotVolumeOption,
    ...
}
```

This way we will disallow having the percentage and USD set at the same time as min_volume, but pct min_volume with USD max_volume will still be valid. I hope it will also make the handling code a bit less complicated.

Reply:
Indeed it's better, it allows to remove the checks of the `is_some`